File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if (!root) {
            return {};
        }

        vector<int> rightMostNodes;

        queue<TreeNode*> nodes;
        nodes.push(root);
        while (!nodes.empty()) {
            for (int nodeIdx = 0, n = nodes.size(); nodeIdx < n; ++nodeIdx) {
                TreeNode* node = nodes.front();
                nodes.pop();
                if (nodeIdx == 0) {
                    rightMostNodes.push_back(node->val);
                }
                if (node->right) {
                    nodes.push(node->right);
                }
                if (node->left) {
                    nodes.push(node->left);
                }
            }
        }
        return rightMostNodes;
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        """
        Find the first node such that 
            the smaller node is under the root and the larger node is greater than the node
        Edge case:
            what if the root is p or q? 
                This test case isn't possible since a lowest common ancestor 
                cannot be the same node as p or q
            duplicate nodes?
                all node values are unique
        Approach one O(N^2):
            for each node, search for the smaller node on the left subtree and
            search for the larger node on the right subtree. if both nodes are found,
            we found our LCA
        Approach two:
            if the root is larger/smaller than both p and q, it cannot be the LCA. Adjust (l or r)

        """

        smaller = q.val
        larger = p.val
        if larger < smaller:
            smaller, larger = larger, smaller
        
        def findLCA(root, smaller, larger):
            if root.val > larger:
                return findLCA(root.left, smaller, larger)
            if root.val < smaller:
                return findLCA(root.right, smaller, larger)
            return root

        return findLCA(root, smaller, larger)